design/open/repository-operations-error-handling.md

```diff
@@ -0,0 +1,121 @@
+
+## Solution
+The proposed solution introduces 2 new elements to solve the above, in terms of DB consistency a.k.a **Correctness**:
+* A `repository_state` that will allow to correctly differ `Repository` in use, from an useless `Repository` due to error, deletion etc
+* A `unique_identifier` for each `Repository` that will allow to differ it from other (previously created and possibly deleted) `Repository` that had the same `RepositoryID`. This will allow to correlate entities to the correct `Repository` entity, in case of `RepositoryID` collision (due to entities left from a previously deleted `Repository` with the same `RepositoryID`)
+In addition:
+* A garbage cleaning mechanism will remove zombie entities, and maintain our KV Store clean and as compact as possible. This GC will identify useless `Repository` entities and will reattempt to delete all the correlated entities, which are now easily recognizable with the addition of the new `unique_identifier`. Note that this GC is not a part of maintaining the DB consistency
+
+### Repository State
+Introduce a new attribute - `repository_state` - to recognize a failed or otherwise irrelevant `Repository` entities. Based on the `state`, a `Repository` that was not completely created, can be identified and treated as `not_exist` as well as a `Repository` in the middle of deletion.
+These unused `Repository` entries, which now can be easily detected, can later be cleaned by a future garbage cleaner, to maintain the KV cleanliness and to reduce its size.
+The new `repository_state` attribute can have one of four values: `initial`, `active`, `deleted` or `failed`
+* Upon creation, a `Repository` entry will have an `initial` state, to indicate it is not usable yet. It will also indicate that this `Repository` is soon to be valid for usage, so other threads/processes treat it as such
+  * A failure in the next creation steps - default `Branch`, initial `Commit` or any other future step, will transfer the `Repository` to state `failed`, if possible. If that cannot be achieved (e.g. the creating process crashed, or KV connectivity is lost) the `Repository` can stay in state `initial` as it is still unused
+  * If during creation, a `Repository` with identical ID (name) is found, the creation of the new `Repository` is stopped before any update the the `KV` is done. This is similar to the current behavior and will be done regardless of the existing `Repository` status
+* When retrieving a `Repository` it will only be retrieved if it is in `active` status. Otherwise an error will be returned.
+  * If the `Repository` entry does not exist, an `ErrRepositoryNotFound` will be returned
```

Comment:
I think we should just return ErrRepositoryNotFound for any state other than active

Reply:
Sounds legit to me.